Fix support search query/result reset functionality

getdave opened the pull request:
This PR fixes an issue in the support search for both:

- the "Get help" card on "My Home"
- the blue "?" FAB inline help

The problem is best summarised in [the Issue](https://github.com/Automattic/wp-calypso/issues/43105):

> When interacting with the new support search card on Calypso My Home, resetting the search query to empty causes the "no results" state to appear in the UI.

## The fix

To fix this I have:

- Fixed to ensure that updating the search _results_ in state also updates the corresponding search _query_ text - otherwise results and query become out of sync!
- ~Add a new action `resetSearchResults` which sets the search results and query to an empty value in the relevant portions of state.~ Not required. Removed.
- Updated implementation of Inline Help (FAB) component to always dispatch the search action even for empty strings. This ensures we get the reset state and contextual results shown. Moreover the components are now almost identical and ripe for the forthcoming refactor!

## Testing instructions

#### On Master

* Boot Calypso and go to "My Home" on any site.
* See the "Get Help" support search card.
* Make a careful note of the initial state of the UI.
* Type a valid search (eg: `Podcast`). See results.
* Clear the search (via delete or hitting the `X`).
* See invalid state whereby "No results" shows and results are not showing the default "contextual" results. 
* _Repeat_ for the FAB and verify same issue.

#### On this PR

* Boot Calypso and go to "My Home" on any site.
* See the "Get Help" support search card.
* Make a careful note of the initial state of the UI.
* Type a valid search (eg: `Podcast`). See results.
* Clear the search (via delete or hitting the `X`).
* See valid state which should be the same as the initial state of the UI before you interacted with it at all. You should see:
  - there is _no_ "No results" message shown.
  - the search results show the "contextual results".
  - the search field is empty.
* Enter invalid search term and repeat the process above to ensure search state can be reset.
* _Repeat_ for the FAB and verify same fix works here.

## Screenshots

![Screen Capture on 2020-06-16 at 12-03-45](https://user-images.githubusercontent.com/444434/84766894-7be34280-afc9-11ea-88aa-2ef78336b264.gif)
![Screen Capture on 2020-06-16 at 12-03-06](https://user-images.githubusercontent.com/444434/84766897-7d146f80-afc9-11ea-82e5-9f2d97d9517b.gif)


Fixes https://github.com/Automattic/wp-calypso/issues/43105

getdave committed:
Fix to ensure results and search query are always in sync

getdave committed:
Reset search results state on UI “reset”

getdave committed:
Creates action to make reset intent clearer

getdave committed:
Fix FAB inline search reset implementation to match Home

marekhrabe reviewed (approved):
This works well for me - no more unexpected leftover states!

getdave commented:
Waiting on @retrofox for final approval as he's doing a deep dive on this.

retrofox commented:
> Waiting on @retrofox for final approval as he's doing a deep dive on this.

Thanks. Sorry, I'm not totally convinced that adding a new action is a good solution. Maybe I'm wrong but I'd like to dive into this. _testing..._

retrofox commented on client/state/inline-help/reducer.js:
```diff
@@ -79,6 +79,7 @@ export const search = withoutPersistence(
 				return {
 					...state,
 					selectedResult: -1,
+					searchQuery: action.searchQuery,
```
I think it will be enough to fix the issue. I think there isn't needed to add extra action.

getdave replied:
It's a good spot. I will work for the "My Home" version as we always dispatch the _action_ we just don't _track_ the search if there is no query (ie: it's empty):

https://github.com/Automattic/wp-calypso/blob/0f42e3ff06c2c40bc2d425c9bf3c9b9e07ea0b2a/client/my-sites/customer-home/cards/features/help-search/search-card.jsx#L91-L102

However, the FAB won't work as we conditionally dispatch the _action_ only when there is a search query:

https://github.com/Automattic/wp-calypso/blob/0f42e3ff06c2c40bc2d425c9bf3c9b9e07ea0b2a/client/blocks/inline-help/inline-help-search-card.jsx#L72-L86

So let's normalise the inline version so it matches the Home version and we'll be good.

This also shows how we need to extract a common set of components to avoid such issues.

I'll do this tomorrow AM.

getdave replied:
OK actually I've just fixed it.

getdave committed:
Fix Inline Search to always dispatch search action even for empty queries

Previously the FAB help didn’t dispatch the search action if the query was empty. This was out of sync with the Home version which did. This led to inconsistent behaviour.

This brings this component into line with the implementation in the My Home search.

Now both components can reset their search state without errors.

getdave commented:
@retrofox I fixed the issue by normalising. Please can you:

1. Retest it works across both FAB and the My Home. It's easy to miss!
2. Double check the tracking is tracking the right events for both.

Much appreciated.

retrofox reviewed (dismissed):
> Double check the tracking is tracking the right events for both.

It removes the _no results_ when the input search is empty. However, it doesn't record search events from the Inline Helper component. :-(

retrofox committed:
inline-help: compose serahc and track

retrofox commented:
@getdave I've [added a commit](https://github.com/Automattic/wp-calypso/pull/43323/commits/eb2611f9a343fb10216b00bf663d95357ff12fd2) that fixes the bug when tracking events in the inline help component. Please feel free to remove it or test it :-D

retrofox reviewed (commented):
removing...

obenland commented on client/blocks/inline-help/inline-help-search-card.jsx:
```diff
@@ -69,40 +65,39 @@ class InlineHelpSearchCard extends Component {
 }
 
+const requestInlineSearchResultsAndTrack = ( searchQuery ) => (
+	! searchQuery || ! ( searchQuery.trim() ).length
```
Could we reverse that logic and make it a positive check to make it a little easier to understand what's happening here?

getdave replied:
Agreed. I've reversed in _both_ locations.

obenland reviewed (approved):
Tested in Calypso live and it fixes the "No results" bug in the My Home search. I couldn't reproduce that happening in the FAB in production.

I think this is good to go once the ternary operator in `requestInlineSearchResultsAndTrack()` was updated

getdave committed:
Improve readability by reversing tracking tenaries in both locations

Addresses https://github.com/Automattic/wp-calypso/pull/43323#discussion_r441152589

getdave commented:
@retrofox Thanks for all your work here on sorting out the tracking. I've tested again and it now looks good to me. Could you give it one more run before I deploy pls?

![Screen Shot 2020-06-17 at 10 00 05](https://user-images.githubusercontent.com/444434/84878216-6df0f880-b081-11ea-94ff-7c1515beb3b2.png)

![Screen Shot 2020-06-17 at 09 59 28](https://user-images.githubusercontent.com/444434/84878220-6f222580-b081-11ea-948b-9f7c7c6c2684.png)

retrofox commented on client/blocks/inline-help/inline-help-search-card.jsx:
```diff
@@ -84,19 +84,18 @@ class InlineHelpSearchCard extends Component {
-		)
-);
+const requestInlineSearchResultsAndTrack = ( searchQuery ) =>
+	searchQuery && searchQuery.trim().length
```
Debatable! 😆

retrofox reviewed (approved):
LGTM. :shipit:

getdave committed:
Remove unused dep

retrofox reviewed (changes requested):
I've found a weird issue in this PR. Let me _Request changes_ before to develop the causes

retrofox commented:
Sometimes when you type quickly defining many queries the component enters in an infinity loop. I could reproduce it just typing, again and again, paying attention at the network tab of the development console. :-(

https://cloudup.com/caxDw4IXIUu

<img src="https://cldup.com/uU03AXSWgl.gif" />

obenland commented:
I also noticed longer response times from the API for search terms that don't map to results, up to 30 seconds at times. It doesn't look like it's caused by this PR though as I can reproduce it in production in the FAB itself, too.

retrofox commented:
> I also noticed longer response times from the API for search terms that don't map to results, up to 30 seconds at times. It doesn't look like it's caused by this PR though as I can reproduce it in production in the FAB itself, too.

This PR introduces this problem for the "Get help" component in the customer home page. I couldn't reproduce it in prod.

retrofox commented:
> I can reproduce it in production in the FAB itself, too.

I can't. Have you paid attention to the requests and how they increase automatically?

![infinity-requests-2](https://user-images.githubusercontent.com/77539/84959849-935c2180-b0d6-11ea-86b9-ea43472899c2.gif)

In the meanwhile...
<img width="169" alt="Screen Shot 2020-06-17 at 8 11 11 PM" src="https://user-images.githubusercontent.com/77539/84959964-e03ff800-b0d6-11ea-9414-f3c4e5b2b12b.png">

obenland commented:
Hm, I didn't notice that before, but I can reproduce that. They do resolve eventually, but the UI doesn't. You're right though, triggering those additional requests seems to be introduced in this PR.

getdave committed:
Remove explicit search reset handler

getdave committed:
Adds new search RESET action type and consequential state update

getdave commented:
@retrofox @obenland I think a regression was introduced when we tried to simplify the PR [here](https://github.com/Automattic/wp-calypso/pull/43323#issuecomment-644782264) when we added the `searchQuery` to the SUCCESS reducer handling.

This caused an infinite loop whereby because the query got updated another request got triggered.

Anyway...I've found a simpler solution and I've [made a PR for that here](https://github.com/Automattic/wp-calypso/pull/43434). It touches fewer files and particularly only 1 component I'd appreciate your review of that instead now.

retrofox commented:
> Anyway...I've found a simpler solution and I've [made a PR for that here](https://github.com/Automattic/wp-calypso/pull/43434). It touches fewer files and particularly only 1 component I'd appreciate your review of that instead now.

👍 Let's prioritize #43434. Thanks!

getdave commented:
Closing in favour of https://github.com/Automattic/wp-calypso/pull/43434
